File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.py
def solution(dartResult):
    score = [0,0,0]
    dart = []
#   먼저 세번의 다트결과를 구분해서 나누기
    currStr = ''
    
    for i in range(len(dartResult)):
        if dartResult[i].isdigit():
            if currStr and currStr[-1].isalpha():
                dart.append(currStr)
                currStr = dartResult[i]
            else:
                currStr += dartResult[i]
            
        elif dartResult[i] == '*' or dartResult[i] == '#':
            currStr += dartResult[i]
            dart.append(currStr)
            currStr = ''
        else:
            currStr += dartResult[i]
            
    if currStr:
        dart.append(currStr)
    
    # S,D,T 에 따라 다른 가중치 부여
    def multipleNum(num,t):
        if t == 'S':
            return num
        elif t == 'D':
            return num * num
        elif t == 'T':
            return num * num * num
    
    #각 회차마다 계산하기 , 가중치 더하기
    for i in range(len(dart)):
        for j in range(len(dart[i])):
            if dart[i][j] == '1' and dart[i][j+1] == '0':
                score[i] += 10
            elif dart[i][j].isdigit() :
                score[i] += int(dart[i][j])
            elif dart[i][j].isalpha():
                score[i] = multipleNum(score[i],dart[i][j])
            elif dart[i][j] == '*':
                if i != 0:
                    score[i-1] *= 2
                score[i] *= 2
            elif dart[i][j] == '#':
                score[i] *= -1
                
    return sum(score)

File: 프로그래머스/1/17682. ［1차］ 다트 게임/［1차］ 다트 게임.js
// 세개의 결과로 분리 -> 숫자가 나오면 분리
// 10을 'C'로 바꾼다음 숫자기준,C기준 분리 후 C를 다시 10으로 변경

const score = {'S':1,'D':2,'T':3}

function solution(dartResult) {
    const result = dartResult.replaceAll('10','C');
    let cur_num = -1;
    let cur_score = '';
    let cur_idx = -1;
    
    const ans = [0,0,0];
    
    for (i of result){
        if (!isNaN(i) || i==='C'){
            cur_num = i
            if (cur_num === 'C'){
                cur_num = 10;
            }
        }else if (i in score){
            cur_score = score[i]
            cur_idx += 1
            ans[cur_idx] = cur_num ** cur_score
            
        }else {
            if (i == '*'){
                ans[cur_idx] = ans[cur_idx] * 2
                if (cur_idx > 0){
                    ans[cur_idx-1] = ans[cur_idx-1] * 2
                }
            }
            else{
                ans[cur_idx] = ans[cur_idx] * -1
            }
        }
    }
    const sum = ans.reduce((acc,curr)=>acc+curr,0)
    return sum
}